### PR TITLE
Add missing documentation for exported functions

### PR DESCRIPTION
**Pull request by shehzan10**
This is probably a very naive version of the docs. I've tried to fill in as much as I could with my knowledge.

I did not add documentation for any functions marked `@private` and did not add for `addDefaults` as that is undergoing restructuring as a part of #223.

Part of #163

### CODE REVIEW

File: lib/PrimitiveHelpers.js [hunk 1]
```diff
@@ -21,6 +21,13 @@ module.exports = {
     transformPrimitives : transformPrimitives
 };
 
+/**
+ * Compare two primitives to check if they share an arrtibute.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: "arrtibute" typo

Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: `attribute accessor` is clearer

File: lib/PrimitiveHelpers.js [hunk 2]
```diff
@@ -63,6 +78,14 @@ function primitivesHaveOverlappingIndexAccessors(gltf, primitive, comparePrimiti
     return false;
 }
 
+/**
+ * Compare two primitives to check if they have an overlapping index accessor.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Fix description.

File: lib/PrimitiveHelpers.js [hunk 3]
```diff
@@ -63,6 +78,14 @@ function primitivesHaveOverlappingIndexAccessors(gltf, primitive, comparePrimiti
...
+ *
+ * @param {Object} gltf A javascript object containing a glTF asset.
+ * @param {Array} primitives An array containing the primitives that need to be transformed.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: `{Object[]}` is preferable

File: lib/PrimitiveHelpers.js [hunk 4]
```diff
@@ -63,6 +78,14 @@ function primitivesHaveOverlappingIndexAccessors(gltf, primitive, comparePrimiti
...
+ * @param {Matrix4} transform The transform to apply to the primitives.
+ * @returns {Boolean} True if primitives have an overlapping index accessor, false otherwise.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Fix description.

File: lib/PrimitiveHelpers.js [hunk 5]
```diff
@@ -163,6 +186,14 @@ function transformPrimitives(gltf, primitives, transform) {
     }
 }
 
+/**
+ * Get a hash table like object with the material as the hash and the
+ * primitives that use the material in an array as the bucket.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: This description can be simplified to `Gets primitives sorted by material.` The return can describe what the object looks like.

File: lib/PrimitiveHelpers.js [hunk 6]
```diff
@@ -163,6 +186,14 @@ function transformPrimitives(gltf, primitives, transform) {
...
+ *
+ * @param {Array} primitives An array containing the primitive to sort.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: `{Object[]}`

File: lib/techniqueAttributeForParameter.js [hunk 7]
```diff
@@ -1,6 +1,13 @@
 'use strict';
 module.exports = techniqueAttributeForParameter;
 
+/**
+ * Retrieves the primitive attributes that has a given parameter name.
+ *
+ * @param {Object} technique A javascript object containing a glTF technique.
+ * @param {String} parameterName The search string for parameter.
+ * @returns {Array.<string>} The primitive attribute with matching paramerterName.
```
Comment (lasalvavida) [inline lib/techniqueAttributeForParameter.js]: Typo: parameterName

File: lib/techniqueAttributeForParameter.js [hunk 8]
```diff
@@ -1,6 +1,13 @@
 'use strict';
 module.exports = techniqueAttributeForParameter;
 
+/**
+ * Retrieves the primitive attributes that has a given parameter name.
```
Comment (lasalvavida) [inline lib/techniqueAttributeForParameter.js]: Not primitive attributes, technique attributes.

File: lib/PrimitiveHelpers.js [hunk 9]
```diff
@@ -163,6 +186,14 @@ function transformPrimitives(gltf, primitives, transform) {
...
+ * @returns {Object} An object with properties names of material types where
+ * each property is an array conmtaining all the primitives that use that material.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Put the doc on one line, which may be required by jsdoc.

Description can be more concise: "Returns an object mapping material ids to an array of primitives"

File: lib/techniqueAttributeForParameter.js (hunk 7 above)
Comment (lasalvavida) [inline lib/techniqueAttributeForParameter.js]: Returns a string, not an array

File: lib/PrimitiveHelpers.js [hunk 10]
```diff
@@ -197,6 +235,12 @@ function getPrimitiveConflicts(primitives, primitive) {
     return conflicts;
 }
 
+/**
+ * Process the glTF asset and return all the primitives in its meshes.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: You can probably drop 'process the glTF asset'

File: lib/PrimitiveHelpers.js [hunk 11]
```diff
@@ -197,6 +235,12 @@ function getPrimitiveConflicts(primitives, primitive) {
...
+ *
+ * @param {Object} gltf a javascript object containing a glTF asset.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: Capitalize 'a'

File: lib/PrimitiveHelpers.js [hunk 12]
```diff
@@ -185,6 +216,13 @@ function getPrimitivesByMaterialMode(primitives) {
     return primitivesByMaterialMode;
 }
 
+/**
+ * Process the glTF asset primitives and return primitives that share attribute accessors.
+ *
+ * @param {Array} primitives An array of primitive objects to compare the primitve against.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: {Object[]}

File: lib/techniqueAttributeForParameter.js (hunk 7 above)
Comment (shehzan10) [inline lib/techniqueAttributeForParameter.js]: This one I got from [here](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/blob/master/lib/getPrimitiveAttributeSemantics.js#L13). Does this need correcting as well?

File: lib/PrimitiveHelpers.js [hunk 13]
```diff
@@ -185,6 +216,13 @@ function getPrimitivesByMaterialMode(primitives) {
     return primitivesByMaterialMode;
 }
 
+/**
+ * Process the glTF asset primitives and return primitives that share attribute accessors.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Remove "Process the glTF asset primitives", no processing happens in here.

File: lib/PrimitiveHelpers.js [hunk 14]
```diff
@@ -185,6 +216,13 @@ function getPrimitivesByMaterialMode(primitives) {
...
+ *
+ * @param {Array} primitives An array of primitive objects to compare the primitve against.
+ * @param {Object} primitive The primitve to compare for conflicts.
+ * @returns {Array} An array conmtaining all the primitives that share attribute accessor.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: {Number[]}

"An array containing indexes of primitives that have attribute accessor conflicts"

File: lib/PrimitiveHelpers.js [hunk 15]
```diff
@@ -163,6 +186,14 @@ function transformPrimitives(gltf, primitives, transform) {
     }
 }
 
+/**
+ * Get a hash table like object with the material as the hash and the
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: There are actually two levels to this, material and mode (e.g. WebGLConstants.TRIANGLES). Also, just call them objects, referring to hash tables here is redundant.

File: lib/PrimitiveHelpers.js (hunk 10 above)
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Remove "Process the glTF asset"

File: lib/PrimitiveHelpers.js [hunk 16]
```diff
@@ -197,6 +235,12 @@ function getPrimitiveConflicts(primitives, primitive) {
     return conflicts;
 }
 
+/**
+ * Process the glTF asset and return all the primitives in its meshes.
+ *
+ * @param {Object} gltf a javascript object containing a glTF asset.
+ * @returns {Array} An array conmtaining all the primitives.
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: {Object[]}

File: lib/PrimitiveHelpers.js (hunk 1 above)
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: Typo. Also, we want to check if they share an attribute accessor/data and that should probably be stated more explicitly.

File: lib/PrimitiveHelpers.js [hunk 17]
```diff
@@ -63,6 +78,14 @@ function primitivesHaveOverlappingIndexAccessors(gltf, primitive, comparePrimiti
     return false;
 }
 
+/**
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: This is copy pasted from `primitivesHaveOverlappingIndexAccessors` but still has the same header and return comment not matching this function.

File: lib/techniqueAttributeForParameter.js (hunk 7 above)
Comment (lilleyse) [inline lib/techniqueAttributeForParameter.js]: This function returns a single string. Change the description and return type.

"paramerterName" -> "parameterName"

File: lib/techniqueParameterForSemantic.js [hunk 18]
```diff
@@ -5,6 +5,13 @@ var defined = Cesium.defined;
 
 module.exports = techniqueParameterForSemantic;
 
+/**
+ * Retrieves the technique parameter that has a matching semantic.
+ *
+ * @param {Object} technique A javascript object containing a glTF technique.
+ * @param {String} semantic The search string for semantics.
+ * @returns {Array.<string>} The technique semantics with matching semantic.
```
Comment (lilleyse) [inline lib/techniqueParameterForSemantic.js]: This return a single String, not an array.

File: lib/techniqueAttributeForParameter.js [hunk 19]
```diff
@@ -1,6 +1,13 @@
...
+ *
+ * @param {Object} technique A javascript object containing a glTF technique.
+ * @param {String} parameterName The search string for parameter.
+ * @returns {Array.<string>} The primitive attribute with matching paramerterName.
+ */
 function techniqueAttributeForParameter(technique, parameterName) {
```
Comment (lilleyse) [inline lib/techniqueAttributeForParameter.js]: In addition can you mark this as `@private`?

File: lib/techniqueParameterForSemantic.js (hunk 18 above)
Comment (lilleyse) [inline lib/techniqueParameterForSemantic.js]: Mark this as `@private` too.

**Comment by lilleyse** [comment]
Ah didn't realize you were also reviewing this @lasalvavida, so there may be duplicates.

File: lib/PrimitiveHelpers.js (hunk 2 above)
Comment (shehzan10) [inline lib/PrimitiveHelpers.js]: Isn't this description correct for this function?

Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Oh, my mistake.

File: lib/PrimitiveHelpers.js [hunk 20]
```diff
@@ -185,6 +213,13 @@ function getPrimitivesByMaterialMode(primitives) {
     return primitivesByMaterialMode;
 }
 
+/**
+ * Return primitives that share attribute accessors with a given primitive.
+ *
+ * @param {Object[]} primitives An array of primitive objects to compare the primitve against.
+ * @param {Object} primitive The primitve to compare for conflicts.
+ * @returns {Number[]} An array conmtaining indices of primitives that have attribute accessor conflicts.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: Typo

File: lib/PrimitiveHelpers.js [hunk 21]
```diff
@@ -163,6 +185,12 @@ function transformPrimitives(gltf, primitives, transform) {
     }
 }
 
+/**
+ * Returns primitives sorted by material.
+ *
+ * @param {Object[]} primitives An array containing the primitive to sort.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: primitives*

File: lib/PrimitiveHelpers.js [hunk 22]
```diff
@@ -163,6 +185,12 @@ function transformPrimitives(gltf, primitives, transform) {
...
+ * @returns {Object} An object mapping material ids to an array of primitives.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: This was not corrected. There are two levels, material -> mode -> primitive[]

File: lib/PrimitiveHelpers.js [hunk 23]
```diff
@@ -197,6 +232,12 @@ function getPrimitiveConflicts(primitives, primitive) {
     return conflicts;
 }
 
+/**
+ * Return all the primitives in its meshes.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: *all the primitives in the meshes of a glTF asset.

File: lib/PrimitiveHelpers.js [hunk 24]
```diff
@@ -197,6 +232,12 @@ function getPrimitiveConflicts(primitives, primitive) {
...
+ *
+ * @param {Object} gltf A javascript object containing a glTF asset.
+ * @returns {Object[]} An array conmtaining all the primitives.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: Typo

File: lib/techniqueAttributeForParameter.js [hunk 25]
```diff
@@ -1,6 +1,15 @@
 'use strict';
 module.exports = techniqueAttributeForParameter;
 
+/**
+ * Retrieves the technique attributes that has a given parameter name.
```
Comment (lasalvavida) [inline lib/techniqueAttributeForParameter.js]: *attribute (singular)

File: lib/techniqueParameterForSemantic.js [hunk 26]
```diff
@@ -5,6 +5,15 @@ var defined = Cesium.defined;
 
 module.exports = techniqueParameterForSemantic;
 
+/**
+ * Retrieves the technique parameter that has a matching semantic.
+ *
+ * @param {Object} technique A javascript object containing a glTF technique.
+ * @param {String} semantic The search string for semantics.
+ * @returns {String} The technique semantics with matching semantic.
```
Comment (lasalvavida) [inline lib/techniqueParameterForSemantic.js]: *technique **parameter** with matching semantic

File: lib/PrimitiveHelpers.js [hunk 27]
```diff
@@ -185,6 +213,13 @@ function getPrimitivesByMaterialMode(primitives) {
     return primitivesByMaterialMode;
 }
 
+/**
+ * Return primitives that share attribute accessors with a given primitive.
+ *
+ * @param {Object[]} primitives An array of primitive objects to compare the primitve against.
+ * @param {Object} primitive The primitve to compare for conflicts.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: Typo: primitve

File: lib/PrimitiveHelpers.js [hunk 28]
```diff
@@ -163,6 +185,12 @@ function transformPrimitives(gltf, primitives, transform) {
     }
 }
 
+/**
+ * Returns primitives sorted by material.
```
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: This needs to be updated as well

**Comment by lasalvavida** [comment]
That's my last comment for this. After you make that change this should be good to merge. Thanks @shehzan10!

**Comment by shehzan10** [comment]
@lasalvavida I pushed the fix for the last commend.

**Comment by lasalvavida** [comment]
Looks good!

**Pull request merged**